jsonrpcservice/__init__.py

```diff
@@ -0,0 +1,259 @@
+    code = -32603
+
+    def __init__(self, message="Internal error", data=None):
+        super().__init__(self.code, message, data)
+
+
+class ApplicationError(JsonRpcError):
+    def __init__(self, code, message="Application error", data=None):
+        if -32000 >= code >= -32768:
+            raise ValueError("Error codes between -32000 and -32768 are reserved by JSON-RPC 2.0 specification.")
+
+        super().__init__(code, message, data)
+
+
+class Response(object, metaclass=abc.ABCMeta):
+    def __init__(self, request):
+        self.request = request
+
+    @abc.abstractmethod
+    def dict(self):
+        pass
+
+    def body(self):
+        if self.request.is_notification():
+            return ""
+        else:
+            return json.dumps(self.dict())
+
+
+class ErrorResponse(Response):
+    def __init__(self, request, code, message, data=None, exception=None):
+        super().__init__(request)
+        self.code = code
+        self.message = message
+        self.data = data
+        self.exception = exception
+
+    def dict(self):
+        error = {
+            'message': self.message,
+            'code': self.code,
+        }
+
+        if self.data is not None:
+            error['data'] = self.data
+
+        return {
+            "jsonrpc": "2.0",
+            "id": self.request.id(),
+            "error": error
+        }
+
+
+class SuccessResponse(Response):
+    def __init__(self, request, result):
+        super().__init__(request)
+        self.result = result
+
+    def dict(self):
+        return {
+            "jsonrpc": "2.0",
+            "id": self.request.id(),
+            "result": self.result
+        }
+
+
+class Request(object):
+    def __init__(self, raw_request=None, parsed_request=None):
+        self._id = None
+        self._method = None
+        self._args = []
+        self._kwargs = {}
+        self._notification = False
+        self._prepared = False
+
+        if raw_request is None and parsed_request is None:
+            raise ValueError('Either "raw_request" or "parsed_request" must be set for Request.')
+
+        self._raw_request = raw_request
+        self._parsed_request = parsed_request
+
+    def parsed_request(self):
+        if self._parsed_request is None:
+            try:
+                self._parsed_request = json.loads(self._raw_request)
+            except (TypeError, ValueError) as e:
+                raise ParseError(message=str(e))
+
+        return self._parsed_request
+
+    def id(self):
+        self._prepare()
+        return self._id
+
+    def method(self):
+        self._prepare()
+        return self._method
+
+    def args(self):
+        self._prepare()
+        return self._args
+
+    def kwargs(self):
+        self._prepare()
+        return self._kwargs
+
+    def is_notification(self):
+        self._prepare()
+        return self._notification
+
+    def _prepare(self):
+        if self._prepared:
+            return
+
+        self._prepared = True
```

Comment:
Setting `prepared` flag here means, that if somebody accesses for example `kwargs` for invalid request, he gets the exception. But when he accesses another computed attribute for the same request, for example `is_notification`, he won't get the exception again, but he gets the default value of `_notification` instead. Is this behaviour intended?

Reply:
It is intended, otherwise you could access no information about the request.

Reply:
It is wonky, but didn't think of anything better yet.

Reply:
And now I did :D Will do it better way.